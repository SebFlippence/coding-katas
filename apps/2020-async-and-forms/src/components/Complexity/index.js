import React, { Component } from "react";
import ReactModal from "react-modal";

import complexOperation from "../../utils";

const modalStyle = {
  content: {
    top: "25%",
    left: "50%",
    right: "auto",
    bottom: "auto",
    marginRight: "-50%",
    transform: "translate(-50%, -50%)"
  }
};

class Complexity extends Component {
  constructor(props) {
    super(props);
    this.state = {
      open: false,
      complexity: 0,
      result: null
    };

    this.openModalHandler = this.openModalHandler.bind(this);
    this.closeModalHandler = this.closeModalHandler.bind(this);
    this.computeHandle = this.computeHandle.bind(this);
    this.handleComplexityChange = this.handleComplexityChange.bind(this);
  }

  componentDidMount() {
    ReactModal.setAppElement(document.querySelector("body"));
  }

  openModalHandler() {
    this.setState({ open: true });
  }

  closeModalHandler() {
    this.setState({
      open: false,
      complexity: 0,
      loading: false
    });
  }

  getComplexValue(complexity) {
    return new Promise(resolve => resolve(complexOperation(complexity)));
  }

  computeHandle() {
    const { complexity } = this.state;
    this.setState({
      open: false,
      loading: true
    });

    this.getComplexValue(complexity).then(result =>
      this.setState({
        open: false,
        loading: true,
        result
      })
    );
  }

  handleComplexityChange(event) {
    this.setState({
      complexity: event.target.value
    });
  }

  showResult() {
    const { result, loading } = this.state;
    if (result) {
      return result;
    }

    if (loading) {
      return "Loading...";
    }

    return "Not computed yet";
  }

  render() {
    const { open, complexity } = this.state;
    const result = this.showResult();
    return (
      <>
        <div className="results-container">
          <div>Value: {result}</div>
          <button type="button" onClick={this.openModalHandler}>
            Compute
          </button>
        </div>
        <ReactModal
          isOpen={open}
          style={modalStyle}
          contentLabel="Compute complex expression"
          onRequestClose={this.closeModalHandler}
        >
          <h1>Complexity</h1>
          <input value={complexity} onChange={this.handleComplexityChange} />
          <button type="button" onClick={this.computeHandle}>
            Apply
          </button>
        </ReactModal>
      </>
    );
  }
}

export default Complexity;
