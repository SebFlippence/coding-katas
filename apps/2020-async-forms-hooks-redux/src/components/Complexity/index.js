import React, { useState, useEffect } from 'react';
import ReactModal from 'react-modal';

import complexOperation from '../../utils';

const modalStyle = {
  content: {
    top: '25%',
    left: '50%',
    right: 'auto',
    bottom: 'auto',
    marginRight: '-50%',
    transform: 'translate(-50%, -50%)',
  },
};

function Complexity() {
  const [state, setState] = useState({
    open: false,
    loading: false,
    complexity: 0,
    result: null,
  });

  useEffect(() => ReactModal.setAppElement(document.querySelector('body')), []);

  const openModalHandler = () => setState({ ...state, open: true });

  const getComplexValue = complexity =>
    new Promise(resolve => setImmediate(() => resolve(complexOperation(complexity))));

  const computeHandle = async () => {
    const { complexity } = state;
    setState({
      ...state,
      open: false,
      loading: true,
    });

    const result = await getComplexValue(complexity);
    setState({
      ...state,
      open: false,
      loading: false,
      result,
    });
  };

  const handleComplexityChange = event =>
    setState({
      ...state,
      complexity: event.target.value,
    });

  const getResult = () => {
    const { result, loading } = state;
    if (loading) {
      return 'Loading...';
    }

    if (result) {
      return result;
    }

    return 'Not computed yet';
  };

  const { open, complexity } = state;
  const result = getResult();
  return (
    <>
      <p>Value: {result}</p>
      <button type="button" onClick={openModalHandler}>
        Compute
      </button>
      <ReactModal isOpen={open} style={modalStyle} contentLabel="Compute complex expression">
        <h1>
          <label htmlFor="complexity">Complexity</label>
        </h1>
        <input id="complexity" value={complexity} onChange={handleComplexityChange} />
        <button type="button" onClick={computeHandle}>
          Apply
        </button>
      </ReactModal>
    </>
  );
}

export default Complexity;
