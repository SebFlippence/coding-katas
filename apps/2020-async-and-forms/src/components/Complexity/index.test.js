import React from 'react';
import { act } from 'react-dom/test-utils';
import { render, fireEvent } from '@testing-library/react';

import Complexity from '.';

describe('Complexity', () => {
  test('rendering correctly', () => {
    const { baseElement } = render(<Complexity />);
    expect(baseElement).toMatchSnapshot();
  });

  test('open modal', () => {
    const { baseElement, getByText } = render(<Complexity />);
    fireEvent.click(getByText('Compute'));
    expect(baseElement).toMatchSnapshot();
  });

  test('enter text and submit', async () => {
    const { baseElement, getByText, getByLabelText } = render(<Complexity />);
    fireEvent.click(getByText('Compute'));
    fireEvent.change(getByLabelText('Complexity'), { target: { value: '1' } })

    await act(async () => {
      await fireEvent.click(getByText('Apply'));
    });

    expect(baseElement).toMatchSnapshot();
  });
});
