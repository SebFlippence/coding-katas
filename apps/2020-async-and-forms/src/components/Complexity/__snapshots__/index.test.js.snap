// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Complexity enter text and submit 1`] = `
<body
  class=""
>
  <div>
    <div
      class="results-container"
    >
      <div>
        Value: 
        209685000000
      </div>
      <button
        type="button"
      >
        Compute
      </button>
    </div>
  </div>
  <div
    class="ReactModalPortal"
  />
</body>
`;

exports[`Complexity open modal 1`] = `
<body
  aria-hidden="true"
  class="ReactModal__Body--open"
>
  <div
    data-react-modal-body-trap=""
    style="position: absolute; opacity: 0;"
    tabindex="0"
  />
  <div>
    <div
      class="results-container"
    >
      <div>
        Value: 
        Not computed yet
      </div>
      <button
        type="button"
      >
        Compute
      </button>
    </div>
  </div>
  <div
    class="ReactModalPortal"
  >
    <div
      class="ReactModal__Overlay ReactModal__Overlay--after-open"
      style="position: fixed; top: 0px; left: 0px; right: 0px; bottom: 0px; background-color: rgba(255, 255, 255, 0.75);"
    >
      <div
        aria-label="Compute complex expression"
        class="ReactModal__Content ReactModal__Content--after-open"
        role="dialog"
        style="position: absolute; top: 25%; left: 50%; border: 1px solid #ccc; background: rgb(255, 255, 255); overflow: auto; border-radius: 4px; outline: none; padding: 20px; margin-right: -50%; transform: translate(-50%, -50%);"
        tabindex="-1"
      >
        <h1>
          <label
            for="complexity"
          >
            Complexity
          </label>
        </h1>
        <input
          id="complexity"
          value="0"
        />
        <button
          type="button"
        >
          Apply
        </button>
      </div>
    </div>
  </div>
  <div
    data-react-modal-body-trap=""
    style="position: absolute; opacity: 0;"
    tabindex="0"
  />
</body>
`;

exports[`Complexity rendering correctly 1`] = `
<body>
  <div>
    <div
      class="results-container"
    >
      <div>
        Value: 
        Not computed yet
      </div>
      <button
        type="button"
      >
        Compute
      </button>
    </div>
  </div>
  <div
    class="ReactModalPortal"
  />
</body>
`;
